services/worker-manager/src/provisioner.js

```diff
@@ -96,26 +104,79 @@ class Provisioner {
    * Run a single provisioning iteration
    */
   async provision() {
+    if (this.provisioningLoopAlive) {
+      this.monitor.notice('loop-interference', {});
+      return;
+    }
+    this.provisioningLoopAlive = true;
     // Any once-per-loop work a provider may want to do
     await this.providers.forAll(p => p.prepare());
 
+    // track the providerIds seen for each worker pool, so they can be removed
+    // from the list of previous provider IDs
+    const providersByPool = new Map();
+    const seen = (providerId, workerPoolId, workerCapacity) => {
+      const v = providersByPool.get(workerPoolId);
+      if (v) {
+        v.providers.add(providerId);
+        v.count += workerCapacity;
+      } else {
+        providersByPool.set(workerPoolId, {providers: new Set([providerId]), count: workerCapacity});
+      }
+    };
+
+    // Check the workers
+    await this.providers.forAll(p => p.scanPrepare());
+    await this.Worker.scan({
+      state: Entity.op.notEqual(this.Worker.states.STOPPED),
+    }, {
+      handler: async worker => {
+        seen(worker.providerId, worker.workerPoolId, worker.capacity);
+        const provider = this.providers.get(worker.providerId);
+        if (provider) {
+          try {
+            await provider.checkWorker({worker});
+          } catch (err) {
+            this.monitor.reportError(err); // Just report it and move on so this doesn't block other providers
```

Comment:
But notably we've already counted its capacity!

One concern here: if checkWorker recognized that the worker is STOPPED, its capacity will still be counted.  That would be fixed on the next iteration, though.  Is that a problem?

Reply:
I think it is not a problem. Just handling it on the next loop is simpler and if the loops are happening quickly enough as we plan, this is fine.